core/typevarcheck.ml

```diff
@@ -71,6 +71,8 @@ let rec is_guarded : TypeVarSet.t -> int -> datatype -> bool =
         | `Application (_, ts) ->
             (* don't treat abstract type constructors as guards *)
             List.for_all (is_guarded_type_arg bound_vars var) ts
+        | `RecursiveApplication { r_args ; _ } ->
```

Comment:
I am now somewhat confused about `typename` (see later comment).  We seem to be overriding it to sometimes mean "type synonym" (alias), and other times more like Haskell's "newtype" or "data".  I think that if we are now introducing generative nominal typing (i.e. ML `datatype` or Haskell `data` that introduces a new type name distinct from all others) --- which is perfectly reasonable as a way of simplifying recursive types! --- then the keywords should reflect this difference.

Reply:
Based on discussion below I think that we still intend that `typename` remains a transparent type synonym (like `type` in Haskell) for now, in which case what Frank wrote above should work (or at least have an informative error message explaining why it is a silly thing to do).

That is, if we are imposing a side-condition that the occurrences of recursively defined types in the RHSs of `typename`s inside of `mutual` blocks all need to be guarded by variant constructors, we should check for this somewhere and have an error message.  I'd be fine with this restriction if it makes life easier for now, though I think in Links things like
```
mu a. (a) -> Int

mu a. (a) -> a
```
are perfectly well defined using `mu` so if we can handle mutual recursion with arbitrary RHSs that would be fine too.

Reply:
Well-spotted. This appears to be a bug in `master` too, since I get the same error with:

```
typename Even(a) = (a) -> Int;

4
```

Reply:
There is a free variable in your program, namely the implicit flexible effect variable.

Reply:
Thanks, @dhil. I have been caught many a time by this sneaky implicit variable.

@frank-emrich -- thinking about your example, this will never yield a `RecursiveApplication` since it's not self-referential.

Perhaps a better example would be:
```
  typename Even(a) = Even(a) -> Int;
```

where `Even` occurs unguarded. I think extending the current guardedness checker, following how it's done for recursive type variables, and adding a guardedness check for each entry in a `Typenames` block would be the best idea.

Reply:
I've looked into this, and I don't believe we need to handle this in any special way. As @dhil clarified, the error that @frank-emrich came across was because of the sneaky implicit effect variable. I can confirm that

```
typename Even(a) = (Even(a)) {}-> Int;
```
is valid.

The possible soundness bugs due to unguardedness are checked in `tests/mutual/unguarded.links` and `tests/mutual/unguarded2.links`.

In 15db7100b0015db7138f7afca86ea5a52ffa0443, I fixed the guardedness and positivity checkers. Since the only place this is invoked is when introducing a (mu-bound) recursive type during type inference (`unify.ml`, `rec_intro_point`), I don't think we need to do anything special here.

I may be missing something though, so I'd appreciate any comments.